host/sysfs/fs_linux.go

```diff
@@ -82,26 +83,33 @@ func (e epollEvent) String() string {
 // One OS thread is needed for all the events. This is more efficient on single
 // core system.
 type eventsListener struct {
-	// file descriptor of the epoll handle itself.
-	epollFd int
-	// pipes to wake up the loop()
-	r, w   *os.File
-	wakeUp <-chan time.Time
+	// Atomic value set to one once fully initialized.
+	initialized int32
 
 	// Mapping of file descriptors to wait on with their corresponding channels.
-	mu  sync.Mutex
+	mu sync.Mutex
+	// File descriptor of the epoll handle itself.
+	epollFd int
+	// Pipes to wake up the EpollWait() system call inside loop().
+	r, w *os.File
+	// Return channel to confirm that EpollWait() was woken up.
+	wakeUp <-chan time.Time
+	// Map of file handles to user listening channel.
 	fds map[int32]chan<- time.Time
 }
 
 // init must be called on a fresh instance.
 func (e *eventsListener) init() error {
-	// In theory this check should be done with a lock but in practice this is
-	// not necessary, as this value is only set with a lock.
-	if e.fds != nil {
+	if atomic.LoadInt32(&e.initialized) != 0 {
 		// Was already initialized.
 		return nil
 	}
 	e.mu.Lock()
+	if atomic.LoadInt32(&e.initialized) != 0 {
```

Comment:
This should also release the lock. If the there are concurrent calls of init  all piled up waiting on the lock, if they get to take the lock it will stop loop().

Reply:
Oh gosh thanks!